platform/darwin/src/local_glyph_rasterizer.mm

```diff
@@ -45,47 +46,72 @@ CTFontDescriptor configuration option (although we'd have to override font size
+            fallbackFontNames = [fallbackFontNames ?: @[] arrayByAddingObjectsFromArray:[@(fontFamily_->c_str()) componentsSeparatedByString:@"\n"]];
+        }
+    }
     
     ~Impl() {
-        if (fontHandle) {
-            CFRelease(fontHandle);
-        }
     }
-
     
-    CTFontRef getFont() {
-        if (!fontFamily) {
-            return NULL;
+    bool isEnabled() { return fallbackFontNames; }
+    
+    CTFontDescriptorRef createFontDescriptor(const FontStack& fontStack) {
+        NSMutableArray *fontNames = [NSMutableArray arrayWithCapacity:fontStack.size() + fallbackFontNames.count];
+        for (auto& fontName : fontStack) {
+            if (fontName != util::LAST_RESORT_ALPHABETIC_FONT && fontName != util::LAST_RESORT_PAN_UNICODE_FONT) {
+                [fontNames addObject:@(fontName.c_str())];
+            }
         }
+        [fontNames addObjectsFromArray:fallbackFontNames];
         
-        if (!fontHandle) {
-          NSDictionary *fontAttributes = @{
-                (NSString *)kCTFontSizeAttribute: [NSNumber numberWithFloat:24.0],
-                (NSString *)kCTFontFamilyNameAttribute: [[NSString alloc] initWithCString:fontFamily->c_str() encoding:NSUTF8StringEncoding]
+        CFMutableArrayRefHandle fontDescriptors(CFArrayCreateMutable(kCFAllocatorDefault, fontNames.count, &kCFTypeArrayCallBacks));
+        for (NSString *name in fontNames) {
+            NSDictionary *fontAttributes = @{
+                (NSString *)kCTFontSizeAttribute: @(util::ONE_EM),
+                (NSString *)kCTFontNameAttribute: name,
+                (NSString *)kCTFontDisplayNameAttribute: name,
+                (NSString *)kCTFontFamilyNameAttribute: name,
             };
-
+            
             CTFontDescriptorRefHandle descriptor(CTFontDescriptorCreateWithAttributes((CFDictionaryRef)fontAttributes));
-            fontHandle = CTFontCreateWithFontDescriptor(*descriptor, 0.0, NULL);
-            if (!fontHandle) {
-                throw std::runtime_error("CTFontCreateWithFontDescriptor failed");
-            }
+            CFArrayAppendValue(*fontDescriptors, *descriptor);
+        }
+
+        CFStringRef keys[] = { kCTFontSizeAttribute,                  kCTFontCascadeListAttribute };
+        CFTypeRef values[] = { (__bridge CFNumberRef)@(util::ONE_EM), *fontDescriptors };
+
+        CFDictionaryRefHandle attributes(
+            CFDictionaryCreate(kCFAllocatorDefault, (const void**)&keys,
+                (const void**)&values, sizeof(keys) / sizeof(keys[0]),
+                &kCFTypeDictionaryKeyCallBacks,
+                &kCFTypeDictionaryValueCallBacks));
+        if (CFArrayGetCount(*fontDescriptors)) {
+            CTFontDescriptorRef firstDescriptor = (CTFontDescriptorRef)CFArrayGetValueAtIndex(*fontDescriptors, 0);
+            return CTFontDescriptorCreateCopyWithAttributes(firstDescriptor, *attributes);
+        } else {
+            return CTFontDescriptorCreateWithAttributes(*attributes);
```

Comment:
Does `fontDescriptors` need to be deallocated at this point?

Reply:
`fontDescriptors` is wrapped in a `CFMutableArrayRefHandle` that releases the `CFFontDescriptorRef` automatically when it goes out of scope. I’d use these handles more extensively in this file, but they’re documented to only work with local variables; passing them around from method to method is tricky enough that I’ve resorted to manual releasing in some cases.

Reply:
As for the `CTFontDescriptorRef` that gets returned: for this file, I’m borrowing the Create rule from Apple’s C APIs, so the expectation is that the caller would be responsible for releasing the return value because the method name contains “create”. But since this is C++, it is very tempting to use C++ pointer wrappers instead of C conventions.

Reply:
I wound up documenting the postcondition and inlining one of the methods that used the Create rule. This file is heavy in Core Foundation and Core Text usage, so any developer hacking on this code would be expected to know the Create and Get rules.